modules/users/server/controllers/users.profile.server.controller.js

```diff
@@ -301,33 +301,64 @@ exports.update = function (req, res) {
 
     // If user is changing email, check if it's available
     function (done) {
-
-      // Check if email changed and proceed with extra checks if so
-      if (req.body.email && req.body.email !== req.user.email) {
-        User.findOne({
-          $or: [
-            { emailTemporary: req.body.email.toLowerCase() },
-            { email: req.body.email.toLowerCase() }
-          ]
-        }, 'emailTemporary email', function (err, emailUser) {
-          // Not free
-          if (emailUser) {
-            // If the user we found with this email is currently authenticated user, let user pass to resend confirmation email
-            if (emailUser._id.equals(req.user._id)) {
-              done(null);
+      var emailChanged = (req.body.email && req.body.email !== req.user.email);
+      var userNameChanged = (req.body.username && req.body.username !== req.user.username);
+
+      if (emailChanged || userNameChanged) {
+        // Check if email changed and proceed with extra checks if so
+        if (emailChanged) {
+          User.findOne({
+            $or: [
+              { emailTemporary: req.body.email.toLowerCase() },
+              { email: req.body.email.toLowerCase() }
+            ]
+          }, 'emailTemporary email', function (err, emailUser) {
+            // Not free
+            if (emailUser) {
+              // If the user we found with this email is currently authenticated user, let user pass to resend confirmation email
+              if (emailUser._id.equals(req.user._id)) {
+                done(null);
+              } else {
+                // Otherwise it was someone else's email. Block the way.
+                return res.status(403).send({
+                  message: 'This email is already in use. Please use another one.'
+                });
+              }
             } else {
-              // Otherwise it was someone else's email. Block the way.
-              return res.status(403).send({
-                message: 'This email is already in use. Please use another one.'
-              });
+              // Free, proceed generating the token
+              done(null);
+            }
+          });
+        }
+        // User wants to change the username
+        if (userNameChanged) {
+          var now = new Date(Date.now());
+          var updatePermission;
+          if (req.user.usernameUpdated) {
+            updatePermission = req.user.usernameUpdated;
+            // username can be changed once in every 3 months
+            updatePermission.setMonth(updatePermission.getMonth() + 3);
+          }
+          if (!req.user.usernameUpdated || (now >= updatePermission)) {
+            if (!req.user.usernameHistory) {
+              req.user.usernameHistory = [];
             }
+            // var updatedInfo = {
```

Comment:
Mongoose gives me an error message saying $pushall is deprecated (and it is)...then I'm not sure if update the library would let us use push. What would be the best way to add an object to an array so mongoose could save it to mongoDb?